### PR TITLE
Downloader implementation

### PR DESCRIPTION
**Pull request by MykolaSolyanko**

### CODE REVIEW

File: src/downloader/downloader.cpp [hunk 1]
```diff
@@ -5,34 +5,178 @@
...
+#include <sys/stat.h>
+
+#include <aos/common/tools/fs.hpp>
+
 #include "downloader.hpp"
+#include "log.hpp"
+
+Downloader::~Downloader()
+{
+    aos::LockGuard lock(mMutex);
```
Comment (al1img) [inline src/downloader/downloader.cpp:18]: Empty lin after lock.

File: src/downloader/downloader.cpp [hunk 2]
```diff
@@ -5,34 +5,178 @@
...
+    mFinishDownload = true;
+    mWaitDownload.NotifyOne();
+}
 
 aos::Error Downloader::Init(DownloadRequesterItf& downloadRequester)
 {
+    LOG_DBG() << "Initialize CM client";
```
Comment (al1img) [inline src/downloader/downloader.cpp:25]: Initialize downloader

File: src/downloader/downloader.cpp [hunk 3]
```diff
@@ -5,34 +5,178 @@
...
+{
+    mFinishDownload = true;
+    mErrProcessImageRequest = err;
+    mWaitDownload.NotifyOne();
+}
+
 aos::Error Downloader::ReceiveImageContentInfo(const ImageContentInfo& content)
 {
-    (void)content;
+    LOG_DBG() << "Receive image content info";
```
Comment (al1img) [inline src/downloader/downloader.cpp:150]: Should be guarded with lock.

File: src/downloader/downloader.cpp [hunk 4]
```diff
@@ -5,34 +5,178 @@
...
+    if (result == nullptr) {
+        LOG_ERR() << "File chunk for unknown file: " << chunk.relativePath;
+
+        SetErrorAndNotify(aos::ErrorEnum::eNotFound);
+
+        return aos::ErrorEnum::eNotFound;
+    }
+
+    if (result->mFile == -1) {
+        aos::StaticString<aos::cFilePathLen> path = aos::FS::JoinPath(mRequestedPath, chunk.relativePath);
```
Comment (al1img) [inline src/downloader/downloader.cpp:86]: could be auto as well

File: tests/downloader/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,22 @@
+# SPDX-License-Identifier: Apache-2.0
+
+cmake_minimum_required(VERSION 3.20.0)
+find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
+
+project(dwonloader_test)
```
Comment (al1img) [inline tests/downloader/CMakeLists.txt:6]: typo `dwonloader...`

File: tests/downloader/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+
+# Aos core configuration
+add_definitions(-include ${APPLICATION_SOURCE_DIR}/../../src/aoscoreconfig.hpp)
+
+# ######################################################################################################################
+# Includes
+# ######################################################################################################################
+target_include_directories(app PRIVATE ${APPLICATION_SOURCE_DIR}/../../src)
```
Comment (al1img) [inline tests/downloader/CMakeLists.txt:14]: Empty line after header

File: tests/downloader/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+target_include_directories(app PRIVATE ${APPLICATION_SOURCE_DIR}/../../../aos_core_lib_cpp/include)
+target_include_directories(app PRIVATE ${CMAKE_CURRENT_BINARY_DIR})
+
+# ######################################################################################################################
+# Target
+# ######################################################################################################################
+target_sources(app PRIVATE
```
Comment (al1img) [inline tests/downloader/CMakeLists.txt:21]: Ditto

File: tests/downloader/src/main.cpp [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+#include <unistd.h>
+
+#include <aos/common/tools/log.hpp>
+#include <aos/common/tools/timer.hpp>
+
+#include "downloader/downloader.hpp"
+
+using namespace aos;
+
+#define UNUSED(x) (void)(x)
```
Comment (al1img) [inline tests/downloader/src/main.cpp:23]: There is macro ARG_UNUSED in zephyr.

File: tests/downloader/src/main.cpp [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+
+static const char* kDownloadPath = "download";
```
Comment (al1img) [inline tests/downloader/src/main.cpp:25]: constexpr?

File: tests/downloader/testcase.yaml [hunk 10]
```diff
@@ -0,0 +1,8 @@
+tests:
+  aoszephyrapp.downloader:
+    tags: downloader
+    timeout: 500
+    integration_platforms:
+      - native_posix
+    platform_allow: native_posix
+
```
Comment (al1img) [inline tests/downloader/testcase.yaml:8]: Remove empty line.

File: src/downloader/downloader.cpp (hunk 1 above)
Comment (MykolaSolyanko) [reply src/downloader/downloader.cpp:18]: Done

File: src/downloader/downloader.cpp (hunk 2 above)
Comment (MykolaSolyanko) [reply src/downloader/downloader.cpp]: Done

File: src/downloader/downloader.cpp (hunk 3 above)
Comment (MykolaSolyanko) [reply src/downloader/downloader.cpp:153]: Done

File: tests/downloader/CMakeLists.txt (hunk 5 above)
Comment (MykolaSolyanko) [reply tests/downloader/CMakeLists.txt]: Done

File: tests/downloader/CMakeLists.txt (hunk 6 above)
Comment (MykolaSolyanko) [reply tests/downloader/CMakeLists.txt:15]: Done

File: tests/downloader/CMakeLists.txt (hunk 7 above)
Comment (MykolaSolyanko) [reply tests/downloader/CMakeLists.txt:23]: Done

File: tests/downloader/src/main.cpp (hunk 8 above)
Comment (MykolaSolyanko) [reply tests/downloader/src/main.cpp]: Remove, because this macros does not uses

File: tests/downloader/testcase.yaml (hunk 10 above)
Comment (MykolaSolyanko) [reply tests/downloader/testcase.yaml]: Done

File: src/logger/logger.cpp [hunk 11]
```diff
@@ -54,6 +54,11 @@ LOG_CALLBACK(cmclient);
 LOG_CALLBACK(launcher);
 LOG_CALLBACK(runner);
 LOG_CALLBACK(storage);
+<<<<<<< HEAD
```
Comment (Eabramov84) [inline src/logger/logger.cpp:57]: merge arifacts

File: tests/downloader/CMakeLists.txt [hunk 12]
```diff
@@ -0,0 +1,24 @@
+# SPDX-License-Identifier: Apache-2.0
```
Comment (Eabramov84) [inline tests/downloader/CMakeLists.txt:1]: you could squash  with downloader implementation commit

File: src/logger/logger.cpp (hunk 11 above)
Comment (MykolaSolyanko) [reply src/logger/logger.cpp]: Done

File: tests/downloader/CMakeLists.txt (hunk 12 above)
Comment (MykolaSolyanko) [reply tests/downloader/CMakeLists.txt:1]: Done

**Review by Eabramov84 — APPROVED** [review]

File: tests/downloader/src/main.cpp [hunk 13]
```diff
@@ -0,0 +1,134 @@
...
+#include <unistd.h>
+
+#include <aos/common/tools/log.hpp>
+#include <aos/common/tools/timer.hpp>
+
+#include "downloader/downloader.hpp"
+
+using namespace aos;
+
+constexpr auto kDownloadPath = "download";
```
Comment (al1img) [inline tests/downloader/src/main.cpp:23]: Use c prefix for const

File: tests/downloader/src/main.cpp [hunk 14]
```diff
@@ -0,0 +1,134 @@
...
+constexpr auto kFileName = "test.txt";
+constexpr auto kDownloadUrl = "http://www.example.com";
+constexpr auto data = "file content";
```
Comment (al1img) [inline tests/downloader/src/main.cpp:26]: cData?

File: tests/downloader/src/main.cpp (hunk 13 above)
Comment (MykolaSolyanko) [reply tests/downloader/src/main.cpp]: Done

File: tests/downloader/src/main.cpp (hunk 14 above)
Comment (MykolaSolyanko) [reply tests/downloader/src/main.cpp]: Done

**Pull request merged**